Add a label to deactivate automatic namespace scheduling

astefanutti opened the pull request:
Currently the namespace controller owns the responsibility of assigning namespaces to physical clusters. The only option for other controllers, like the `deployment-splitter` one, or users manually, to distribute resources across physical clusters is to deactivate the namespace controller by starting KCP with `--run-controllers=false --unsupported-run-individual-controllers="workspace-scheduler,cluster"`.

Meanwhile more advanced scheduling / placement scenarios are supported, this PR proposes to introduce the `experimental.scheduling.kcp.dev/disabled` label, that can be added to namespaces, in order to deactivate the automatic placement to physical clusters, performed by the namespace controller.

Setting this label to `false` prevents the namespace controller from overriding the scheduling logic performed by the `deployment-splitter` controller, but also manual scheduling, when working on global load-balancing use cases for example.

Fixes #494

ncdc commented:
Thanks for the PR. Is this for #494?

ncdc commented:
cc @marun @imjasonh

astefanutti commented:
> Thanks for the PR. Is this for #494?

Ah I missed it, It looks like exactly for it, only the name of the label differs, naming is the hardest... :)

marun commented on pkg/reconciler/namespace/namespace.go:
```diff
@@ -268,6 +275,15 @@ func (c *Controller) reconcileNamespace(ctx context.Context, lclusterName string
+	if ns.Labels[clusterLabel] == "" && !scheduleRequirement.Matches(labels.Set(ns.Labels)) {
+		klog.Infof("Skipping placement for namespace %q|%q", lclusterName, ns.Name)
+		return c.patchStatus(ctx, ns, func(conditionsSetter conditions.Setter) {
+			conditions.MarkFalse(conditionsSetter, NamespaceScheduled, NamespaceReasonUnschedulable,
```
Maybe a new status reason is suggested to differentiate 'unable to schedule' from 'scheduling disabled'?

astefanutti replied:
I agree, I've pushed 395fc13886175bcda5e5a1be685ba7729fbd5c28 (I'll squash once the review is complete), that adds the the `SchedulingDisabled` condition reason.

It has actually highlighted an issue in the condition update logic, where the condition was not created until the namespace was actually scheduled. It's fixed with 395fc13886175bcda5e5a1be685ba7729fbd5c28.

marun commented on pkg/reconciler/namespace/namespace.go:
```diff
@@ -190,11 +196,11 @@ func (c *Controller) assignCluster(ctx context.Context, ns *corev1.Namespace) er
 	if oldClusterName != newClusterName {
-		klog.Infof("Patching to update cluster assignment for namespace %q|%q: %q -> %q", ns.ClusterName, ns.Name, oldClusterName, newClusterName)
-		if _, err := c.kubeClient.Cluster(ns.ClusterName).CoreV1().Namespaces().Patch(ctx, ns.Name, types.MergePatchType, clusterLabelPatchBytes(newClusterName), metav1.PatchOptions{}); err != nil {
+		klog.Infof("Patching to update scheduled cluster for namespace %q|%q: %q -> %q", ns.ClusterName, ns.Name, oldClusterName, newClusterName)
```
Why do you propose to update 'cluster assignment' to 'scheduled cluster' everywhere? A resource is scheduled to a cluster. A resource has a cluster assignment/a cluster assigned to it. A resource doesn't have a scheduled cluster.

astefanutti replied:
Agreed, I've reverted the change. I think that was consistency gone wrong :)

marun commented on pkg/reconciler/namespace/namespace.go:
```diff
@@ -268,6 +275,15 @@ func (c *Controller) reconcileNamespace(ctx context.Context, lclusterName string
 		ns.Labels = map[string]string{}
 	}
 
+	if ns.Labels[clusterLabel] == "" && !scheduleRequirement.Matches(labels.Set(ns.Labels)) {
```
I think it may be potentially confusing that the logic for disabling scheduling for a namespace is both here and in `ensureScheduled`. Maybe localize the change to simplify maintenance? e.g.

```golang
schedulingDisabled := !scheduleRequirement.Matches(labels.Set(ns.Labels))
if schedulingDisabled {
  return c.ensureSchedulingDisabledStatus(...)
}
```

I'm also not sure why it makes sense to check for an empty cluster assignment. Once a namespace has been labeled with `kcp.dev/schedule=false`, why should the controller consider the value of `kcp.dev/cluster` for that namespace?

astefanutti replied:
Totally agree. I've updated the logic in 00b0aa88bc13e64e2ea8914ef6ad927ce220f5e2.

marun commented on pkg/reconciler/namespace/namespace.go:
```diff
@@ -37,28 +38,28 @@ import (
-	unscheduledSelector, err = labels.Parse(unscheduled)
-	if err != nil {
-		klog.Fatalf("error parsing label selector %q: %v", unscheduled, err)
+	// This matches namespaces that haven't been scheduled yet
```
Why does it make sense to make this check more restrictive (label does not exist) than previous (label does not exist or does exist and has a non-empty value)? 

While I can see that your intent is that the controller remove the label, if the label were set to an empty value (e.g. `kcp.dev/cluster=`), the code as proposed would appear neither to ensure removal of the label or ensure queueing of the namespace when cluster readiness status changed.

astefanutti replied:
My understanding is that the existing code has to deal with `kcp.dev/cluster=`, because the logic that updates the label does not handle the removal of the label, and simply sets it to `""`. So my rationale was that with properly handling the label deletion, there is no need to deal with `kcp.dev/cluster=` anymore.

marun replied:
I'm afraid it's not an option to assume that the label in question will only be managed by this controller. There is always the potential for other actors - including humans - to make changes that don't match this controller's expectations.  Not assuming anything about the input is also a good way to ensure forwards and backwards compatibility.

marun replied:
bump

astefanutti replied:
True. I've restore the existing logic.

sttts commented:
Am uncomfortable with adding features like this without a bigger picture of direction. Adding a label to control who schedules a namespace is clearly possible, but doesn't add much deeper insights in what we are building.

Can we turn this into some exploration doc about how deployment-splitter like behaviour fits into the bigger TMC picture?

astefanutti commented:
> Am uncomfortable with adding features like this without a bigger picture of direction. Adding a label to control who schedules a namespace is clearly possible, but doesn't add much deeper insights in what we are building.
> 
> Can we turn this into some exploration doc about how deployment-splitter like behaviour fits into the bigger TMC picture?

I agree, this is also my intuition as I've been working through this PR. It seems some exploration is required to address the scheduling / placement concern holistically, possibly looking at https://github.com/open-cluster-management-io/placement, and OCM in general.

On the other hand, the namespace controller, as it stands, pre-empts the scheduling, so the only solution to demo use cases that distribute namespace resources across clusters, is to deactivate it. So maybe fixing #494 in the short term with this PR and do the exploration work in the medium term is an acceptable tradeoff?

astefanutti commented on test/e2e/reconciler/namespace/controller_test.go:
```diff
@@ -28,21 +28,22 @@ import (
 const clusterLabel = "kcp.dev/cluster"
 
 func TestNamespaceScheduler(t *testing.T) {
-	t.Parallel()
```
The namespace controller relies on the physical clusters created in the test fixture, and now that there are more than one test case, that manipulate that "shared state", that leads to race conditions. That explains why I've deactivated parallel execution of the test cases, as I wasn't sure what's the best approach. To isolate that clusters "state", each test case could use a dedicated workspace, but I went for the path of least resistance, so as to get your feedback before proceeding.

marun replied:
I think it would be preferable for now to move the fixture instantiation inside `t.Run()` to ensure that each test gets clean fixture.

astefanutti replied:
Agreed. I've moved the fixture instantiation where you suggested.

astefanutti commented on test/e2e/reconciler/namespace/controller_test.go:
```diff
@@ -73,13 +74,148 @@ func TestNamespaceScheduler(t *testing.T) {
+
+				t.Log("Create a namespace with manual scheduling and expect it to be marked unscheduled")
+
+				namespace, err := server.client.CoreV1().Namespaces().Create(ctx, &corev1.Namespace{
```
I hesitated on that also :) That new "test case" can definitely be added to the existing one. I went for the "single responsibility" principle (if such thing applies to test cases :)), to avoid going down the road of writing a giant test case, but that may have been a premature decision.

That being said, my understanding is that the biggest contributor to total execution time is the creation and tear down of the servers, and adding an extra test case is negligible.

marun replied:
I think that's fine for now. Note that corner-cases will ideally be validated in unit tests as per #559.

astefanutti commented on pkg/reconciler/namespace/namespace.go:
```diff
@@ -97,10 +98,15 @@ func (c *Controller) reconcileResource(ctx context.Context, lclusterName string,
 		return nil
 	}
+	if new == "" {
+		// The namespace hasn't been assigned a cluster.
```
Thinking about it more, that was only an approximation of what is probably a more general solution, that is, do not schedule resources transitively for namespaces with `kcp.dev/schedule=false`. That label means, the namespace controller should not be responsible for the scheduling of that namespace, nor its resources, so the responsibility is fully delegated to an external controller or the user.

I've amended the logic based on that rationale in b386ec6b88652bac5d1dbe720824bd624285ef99.

astefanutti commented:
@marun thanks a lot for the review. I should have covered all your comments, if you want to give a second iteration on this.

ncdc commented:
> Can we turn this into some exploration doc about how deployment-splitter like behaviour fits into the bigger TMC picture?

#557 

> On the other hand, the namespace controller, as it stands, pre-empts the scheduling, so the only solution to demo use cases that distribute namespace resources across clusters, is to deactivate it. So maybe fixing https://github.com/kcp-dev/kcp/issues/494 in the short term with this PR and do the exploration work in the medium term is an acceptable tradeoff?

I chatted with @sttts and we are 👍 on this plan. The label in this PR might not survive as-is for the long-term, but we will ensure that things like the deployment-splitter can coexist with namespace scheduling.

ncdc commented:
Would you want to also check for the label per resource, so that a namespace could be scheduled to a cluster, but you could opt out individual e.g. deployments?

ncdc commented:
@marun assigned to you 😄

sttts commented:
Please some clear indication that the label is experimental and going away. experimental.something.kcp.dev.

marun commented:
> Would you want to also check for the label per resource, so that a namespace could be scheduled to a cluster, but you could opt out individual e.g. deployments?

We had a use case for skipping entire namespaces as per #494. Do we also have one or more documented use cases for skipping individual resources? afaik the goal for this PR is allowing scheduling controllers (e.g. deployment-splitter) to manage specific types vs individual resources.

marun commented on pkg/reconciler/namespace/namespace.go:
```diff
@@ -98,9 +106,10 @@ func (c *Controller) reconcileResource(ctx context.Context, lclusterName string,
 		return nil
 	}
 
-	// Update the resource's assignment.
```
nit: revert to 'assignment' from 'scheduled cluster'

astefanutti replied:
Reverted.

marun commented on pkg/reconciler/namespace/namespace.go:
```diff
@@ -191,10 +209,10 @@ func (c *Controller) assignCluster(ctx context.Context, ns *corev1.Namespace) er
 	if oldClusterName != newClusterName {
 		klog.Infof("Patching to update cluster assignment for namespace %q|%q: %q -> %q", ns.ClusterName, ns.Name, oldClusterName, newClusterName)
-		if _, err := c.kubeClient.Cluster(ns.ClusterName).CoreV1().Namespaces().Patch(ctx, ns.Name, types.MergePatchType, clusterLabelPatchBytes(newClusterName), metav1.PatchOptions{}); err != nil {
+		t, patch := clusterLabelPatchBytes(newClusterName)
```
nit: Please avoid using single-char variables except for receivers and loop indexes.

astefanutti replied:
I've renamed the variables.

marun commented on pkg/reconciler/namespace/namespace.go:
```diff
@@ -236,24 +254,34 @@ func (c *Controller) ensureScheduled(ctx context.Context, lclusterName string, n
+	disabled := !scheduleRequirement.Matches(labels.Set(ns.Labels))
+
+	if condition := conditions.Get(&NamespaceConditionsAdapter{ns}, NamespaceScheduled); condition != nil {
+		conditionStatus := scheduled != (condition.Status == corev1.ConditionTrue)
```
nit: `s/conditionStatus/conditionStatusChanged/` ? i.e. indicate clearly in the name what the conditional indicates. Same comment for `conditionReason` below

astefanutti replied:
Agreed. I've updated both variables as per your suggestion.

marun commented on test/e2e/reconciler/namespace/controller_test.go:
```diff
@@ -107,94 +244,43 @@ func TestNamespaceScheduler(t *testing.T) {
 func scheduledMatcher(target string) namespaceExpectation {
 	return func(object *corev1.Namespace) error {
-		if !nscontroller.IsScheduled(object) {
+		if scheduled := nscontroller.IsScheduled(object); !scheduled {
```
nit: This can be reverted

astefanutti replied:
Thanks for the catch! I had reverted half-way, now fully reverted.

marun commented on test/e2e/reconciler/namespace/controller_test.go:
```diff
@@ -230,3 +316,14 @@ func expectNamespaces(ctx context.Context, t *testing.T, client kubernetes.Inter
 	}, nil
 }
+
+func patchNamespaceLabel(label, value string) []byte {
```
Maybe just make this public in the target package so it can be reused?

astefanutti replied:
I've made the function in the target package and removed that one.

astefanutti commented:
@marun thanks a lot for that second iteration. I should have addressed all your current comments, so it's available for another review cycle, and squashed into a single commit.

@sttts I've changed the label to `experimental.scheduling.kcp.dev/disabled`, and turned it into a _flag_ label, to remove ambiguity on the value.

marun commented:
Please rebase. And I'll be rebasing #559 on this PR to avoid causing you further pain.

astefanutti commented:
@marun thanks for the heads-up. Rebased 👍🏼.

marun commented:
I'm afraid #508 will need to land ahead of this, and I'll do the final review after the inevitable rebase.

astefanutti commented:
@marun thanks, again :) Rebased 👍🏼.

imjasonh commented on pkg/reconciler/namespace/namespace.go:
```diff
@@ -323,10 +370,16 @@ func (c *Controller) reconcileNamespace(ctx context.Context, lclusterName string
-	return []byte(fmt.Sprintf(`
+// ClusterLabelPatchBytes returns JSON patch bytes expressing an operation to
+// add, replace to the given value, or delete the cluster assignment label.
+func ClusterLabelPatchBytes(val string) (types.PatchType, []byte) {
```
This seems to only be exported so that it can be called from e2e tests. If that's the case, I might suggest either reimplementing necessary bits in the test itself (i.e., test that the patch is `"[{"op":...}]` instead of reusing the helper), or perhaps moving the method to an `/internal/` package so that it can't be depended on outside of this Go module.

Exported methods have a nasty habit of getting depended on outside of our codebase, and sometimes beyond intended usage, which can make changes harder to make safely.

astefanutti replied:
I agree. @marun what do you think of moving that function in the `/internal/` package? Could that be an acceptable compromise, as you raised concerns about duplicating it?

marun replied:
Agreed that this detail shouldn't be exposed. I don't think jumping through the hoops of moving to internal is suggested though. How about we avoid using low-level patch semantics in testing and rely on `retry.RetryOnConflict` instead?

e.g. https://github.com/kcp-dev/kcp/blob/main/test/e2e/reconciler/ingress/controller_test.go#L102

sttts commented on pkg/reconciler/namespace/namespace.go:
```diff
@@ -132,25 +151,35 @@ func (c *Controller) patchStatus(ctx context.Context, ns *corev1.Namespace, upda
 	}
-	_, err = c.kubeClient.Cluster(ns.ClusterName).CoreV1().Namespaces().Patch(ctx, ns.Name, types.MergePatchType, patchBytes, metav1.PatchOptions{}, "status")
+	_, err = c.kubeClient.Cluster(ns.ClusterName).CoreV1().Namespaces().
+		Patch(ctx, ns.Name, types.MergePatchType, patchBytes, metav1.PatchOptions{}, "status")
```
please split features from refactorings

sttts commented on pkg/reconciler/namespace/namespace.go:
```diff
@@ -132,25 +151,35 @@ func (c *Controller) patchStatus(ctx context.Context, ns *corev1.Namespace, upda
 // conditions.
-func (c *Controller) updateSchedulableCondition(ctx context.Context, ns *corev1.Namespace, schedulable bool) error {
-	klog.Infof("Patching namespace %q|%q status to indicate schedulable=%t", ns.ClusterName, ns.Name, schedulable)
+func (c *Controller) updateScheduledCondition(ctx context.Context, ns *corev1.Namespace, scheduled, disabled bool) error {
```
please split features from refactorings

sttts commented:
As a general rule of thumb: we should try to split refactorings/cleanups from actual feature changes better. E.g. if we rename a func, a condition, or anything with lots of ripples: do that in its own commit.

marun commented:
There are some nits remaining (including not exporting the patch bytes method just for testing) that I will address in a refactor/cleanup I am working on (of which #559 is a part).
